modules/tinymce/src/core/main/ts/textpatterns/keyboard/Keyboard.ts

```diff
@@ -9,28 +9,50 @@ const setup = (editor: Editor): void => {
   const charCodes = [ ',', '.', ';', ':', '!', '?' ];
   const keyCodes = [ 32 ];
 
-  const getPatternSet = () => Pattern.createPatternSet(Options.getTextPatterns(editor));
-  const getInlinePatterns = () => Pattern.getInlinePatterns(Options.getTextPatterns(editor));
+  // This is a thunk so that they reflect changes in the underlying options each time they are requested.
+  const getPatternSet = () => Pattern.createPatternSet(
+    Options.getTextPatterns(editor),
+    Options.getTextPatternsLookup(editor)
+  );
+
+  // Only used for skipping text pattern matching altogether if nothing has been defined.
+  const hasDynamicPatterns = () => Options.hasTextPatternsLookup(editor);
 
   editor.on('keydown', (e) => {
     if (e.keyCode === 13 && !VK.modifierPressed(e)) {
-      if (KeyHandler.handleEnter(editor, getPatternSet())) {
+      const patternSet = getPatternSet();
+      // Do not process anything if we don't have any inline patterns, block patterns,
+      // or dynamic lookup defined
+      const hasPatterns = patternSet.inlinePatterns.length > 0 ||
+        patternSet.blockPatterns.length > 0 ||
+        hasDynamicPatterns();
+
+      if (editor.selection.isCollapsed() && hasPatterns && KeyHandler.handleEnter(editor, patternSet)) {
```

Comment:
This still eagerly evaluates the patterns which we don't want if the selection is collapsed.